tokio-timer/src/delay_queue.rs

```diff
@@ -507,24 +513,27 @@ impl<T> DelayQueue<T> {
         // Normalize the deadline. Values cannot be set to expire in the past.
         let when = self.normalize_deadline(when);
 
-        // This is needed only for the debug assertion inside the if-let.
-        let old = self.start + Duration::from_millis(self.slab[key.index].when);
-
         self.slab[key.index].when = when;
+        self.insert_idx(when, key.index);
 
-        if let Some(ref mut delay) = self.delay {
-            debug_assert!(old >= delay.deadline());
-
-            let start = self.start;
-            let next_poll = self.wheel.poll_at()
-                .map(move |t| start + Duration::from_millis(t));
-
-            if next_poll != Some(delay.deadline()) {
-                delay.reset(self.start + Duration::from_millis(when));
+        if let Some((index, ref mut delay)) = self.first_to_exp {
+
+            // We reset the delay of the next item to expire only if:
+            //
+            // 1. The new delay is sooner than the current one
+            //
+            // 2. The item to expire first is the same as the one
+            // we are resetting AND the instant to which we are resetting
+            // is further out in time...
+
+            let new_exp = self.start + Duration::from_millis(when);
+            let reset_item_expires_first = new_exp < delay.deadline();
+            let reset_item_is_current_first_exp = index == key.index;
+            let reset_to_later_exp = new_exp > delay.deadline();
+            if reset_item_expires_first || (reset_item_is_current_first_exp && reset_to_later_exp) {
+                delay.reset(new_exp);
```

Comment:
If there are two elements in the queue, and the initially-sooner one is reset to be later than the other, won't this result in the delay being set incorrectly? Perhaps `new_exp` should be taken from `wheel.poll_at()`?

Reply:
Yes.. you are right. And in this case we need to be able to be able to exchange the key as well to be the key of the new "about to expire" item..  Hmm this tracking of first to expire + resetting the timeouts introduces some headache...

Reply:
Will have to give it a bit more thought to make sure I cover all cases